src/backends/postgresql/session.cpp

```diff
@@ -29,6 +30,128 @@ void hard_exec(postgresql_session_backend & session_backend,
+    if (error_code > 0)
+    {
+        len_esc = 0;
+    }
+    retv[len_esc + 1] = '\'';
+    retv.resize(len_esc + 2);
+
+    return retv;
+}
+
+// helper function to collect schemas from search_path
+std::vector<std::string> get_schema_names(postgresql_session_backend & session, PGconn * conn)
+{
+    std::vector<std::string> schema_names;
+    postgresql_result search_path_result(session, PQexec(conn, "SHOW search_path"));
+    if (search_path_result.check_for_data("search_path doesn't exist"))
+    {
+        if (PQntuples(search_path_result) > 0)
+        {
+            std::string search_path_content = PQgetvalue(search_path_result, 0, 0);
+
+            bool quoted = false;
+            std::string schema;
+            while (!search_path_content.empty())
+            {
+                switch (search_path_content[0])
+                {
+                case '"':
+                    quoted = !quoted;
+                    break;
+                case ',':
+                case ' ':
+                    if (!quoted)
+                    {
+                        if (search_path_content[0] == ',')
+                        {
+                            schema_names.push_back(schema);
+                            schema = "";
+                        }
+                        break;
+                    }
+                    SOCI_FALLTHROUGH;
+                default:
+                    schema.push_back(search_path_content[0]);
+                }
+                search_path_content.erase(search_path_content.begin());
+            }
+            if (!schema.empty())
+                schema_names.push_back(schema);
+            for (std::string& schema_name: schema_names)
+            {
+                if (schema_name == "$user")
+                {
+                    postgresql_result current_user_result(session, PQexec(conn, "SELECT current_user"));
+                    if (current_user_result.check_for_data("current_user is not defined"))
+                    {
+                        if (PQntuples(current_user_result) > 0)
+                        {
+                            schema_name = PQgetvalue(current_user_result, 0, 0);
+                        }
+                    }
+                }
+
+                // Ensure no bad characters
+                schema_name = quote(conn, schema_name);
+            }
+        }
+    }
+    if (schema_names.empty())
+    {
+        postgresql_result current_user_result(session, PQexec(conn, "SELECT current_user"));
+        if (current_user_result.check_for_data("current_user is not defined"))
+        {
+            if (PQntuples(current_user_result) > 0)
+            {
+                std::string user = PQgetvalue(current_user_result, 0, 0);
+
+                // Ensure no bad characters
+                schema_names.push_back(quote(conn, user));
+            }
+        }
+        schema_names.push_back("public");
+    }
```

Comment:
I.e. this duplicated code wouldn't be needed if we just set `search_path_content` to the default value if we couldn't get it from the database.